File: DevExpress.OutlookInspiredApp.Wpf/Common/View/Services/ReportService.cs
using System;
using System.ComponentModel;
using System.Printing;
using System.Windows;
using DevExpress.DevAV.Common.ViewModel;
using DevExpress.Mvvm.UI;
using DevExpress.Xpf.Printing;
using DevExpress.Xpf.Ribbon;
using DevExpress.XtraReports;

namespace DevExpress.DevAV.Common.View {
    public abstract class ReportServiceBase : ServiceBase, IReportService {
        class PrintDirectXtraReportPreviewModel : XtraReportPreviewModel {
            protected override void PrintDirect(object parameter) {
                if(parameter is string)
                    PrintDirect((string)parameter);
                else if(parameter is PrintQueue)
                    PrintDirect((PrintQueue)parameter);
                else
                    PrintDirect();
            }
        }

        bool isVisible;
        IReportInfo defaultReportInfo;
        IReportInfo reportInfo;
        IReportInfo actualReportInfo;

        protected bool IsVisible {
            get { return isVisible; }
            set {
                isVisible = value;
                UpdateReport();
                if(!isVisible)
                    this.reportInfo = null;
            }
        }
        protected virtual void SetDefaultReport(IReportInfo reportInfo) {
            this.defaultReportInfo = reportInfo;
            UpdateReport();
        }
        protected virtual void ShowReport(IReportInfo reportInfo) {
            this.reportInfo = reportInfo;
            UpdateReport();
        }
        void UpdateReport() {
            UpdateReportCore(IsVisible ? (reportInfo ?? defaultReportInfo) : null);
        }
        protected virtual void UpdateReportCore(IReportInfo actualReportInfo) {
            UnsubscribeFromParametersViewModel();
            this.actualReportInfo = actualReportInfo;
            SubscribeToParametersViewModel();
            if(this.actualReportInfo == null)
                DestroyReport();
            else
                CreateReport();
        }
        void OnParametersViewModelPropertyChanged(object sender, PropertyChangedEventArgs e) {
            CreateReport();
        }
        void CreateReport() {
            IReport report = actualReportInfo.CreateReport();
            PrintDirectXtraReportPreviewModel model = GetPreviewModel() as PrintDirectXtraReportPreviewModel;
            if(model == null) {
                model = new PrintDirectXtraReportPreviewModel() { ZoomMode = new ZoomFitModeItem(ZoomFitMode.PageWidth) };
                SetPreviewModel(model);
            }
            model.Report = report;
            SetCustomSettingsViewModel(actualReportInfo.ParametersViewModel);
            report.PrintingSystemBase.ClearContent();
            report.CreateDocument(true);
        }
        void DestroyReport() {
            SetCustomSettingsViewModel(null);
            SetPreviewModel(null);
        }
        protected abstract IPreviewModel GetPreviewModel();
        protected abstract void SetPreviewModel(IDocumentPreviewModel previewModel);
        protected abstract void SetCustomSettingsViewModel(object customSettingsViewModel);
        object ActualParametersViewModel { get { return this.actualReportInfo == null ? null : this.actualReportInfo.ParametersViewModel; } }
        void SubscribeToParametersViewModel() {
            INotifyPropertyChanged parametersViewModel = ActualParametersViewModel as INotifyPropertyChanged;
            if(parametersViewModel != null)
                parametersViewModel.PropertyChanged += OnParametersViewModelPropertyChanged;
        }
        void UnsubscribeFromParametersViewModel() {
            INotifyPropertyChanged parametersViewModel = ActualParametersViewModel as INotifyPropertyChanged;
            if(parametersViewModel != null)
                parametersViewModel.PropertyChanged -= OnParametersViewModelPropertyChanged;
        }
        #region IReportService
        void IReportService.SetDefaultReport(IReportInfo reportInfo) {
            SetDefaultReport(reportInfo);
        }
        void IReportService.ShowReport(IReportInfo reportInfo) {
            ShowReport(reportInfo);
        }
        #endregion
    }
    public class BackstageDocumentPreviewReportService : ReportServiceBase {
        public static readonly DependencyProperty BackstageViewIsOpenProperty =
            DependencyProperty.Register("BackstageViewIsOpen", typeof(bool), typeof(BackstageDocumentPreviewReportService), new PropertyMetadata(false, (d, e) =>
                ((BackstageDocumentPreviewReportService)d).OnBackstageViewIsOpenChanged()));
        public static readonly DependencyProperty BackstageItemProperty =
            DependencyProperty.Register("BackstageItem", typeof(BackstageTabItem), typeof(BackstageDocumentPreviewReportService), new PropertyMetadata(null, (d, e) =>
                ((BackstageDocumentPreviewReportService)d).OnBackstageItemChanged((BackstageItem)e.OldValue, (BackstageItem)e.NewValue)));
        public static readonly DependencyProperty BackstageDocumentPreviewProperty =
            DependencyProperty.Register("BackstageDocumentPreview", typeof(BackstageDocumentPreview), typeof(BackstageDocumentPreviewReportService));

        public bool BackstageViewIsOpen {
            get { return (bool)GetValue(BackstageViewIsOpenProperty); }
            set { SetValue(BackstageViewIsOpenProperty, value); }
        }
        public BackstageTabItem BackstageItem {
            get { return (BackstageTabItem)GetValue(BackstageItemProperty); }
            set { SetValue(BackstageItemProperty, value); }
        }
        public BackstageDocumentPreview BackstageDocumentPreview {
            get { return (BackstageDocumentPreview)GetValue(BackstageDocumentPreviewProperty); }
            set { SetValue(BackstageDocumentPreviewProperty, value); }
        }

        protected override void ShowReport(IReportInfo reportInfo) {
            base.ShowReport(reportInfo);
            Dispatcher.BeginInvoke((Action)OpenBackstageView);
        }
        void OpenBackstageView() {
            BackstageItem.Backstage.SelectedTab = BackstageItem;
            BackstageItem.Backstage.IsOpen = true;
        }
        protected override void UpdateReportCore(IReportInfo actualReportInfo) {
            base.UpdateReportCore(actualReportInfo);
            BackstageItem.IsEnabled = actualReportInfo != null;
        }
        void OnBackstageViewIsOpenChanged() {
            IsVisible = BackstageViewIsOpen;
        }
        void OnBackstageItemChanged(BackstageItem oldItem, BackstageItem newItem) {
            if(newItem != null)
                newItem.IsEnabled = oldItem != null && oldItem.IsEnabled;
        }
        protected override IPreviewModel GetPreviewModel() {
            return BackstageDocumentPreview.Model;
        }
        protected override void SetPreviewModel(IDocumentPreviewModel previewModel) {
            BackstageDocumentPreview.Model = previewModel;
        }
        protected override void SetCustomSettingsViewModel(object customSettingsViewModel) {
            BackstageDocumentPreview.CustomSettings = customSettingsViewModel;
        }
    }
    public class DocumentViewerReportService : ReportServiceBase {
        DocumentViewer DocumentViewer { get { return (DocumentViewer)AssociatedObject; } }

        protected override void OnAttached() {
            base.OnAttached();
            IsVisible = true;
        }
        protected override void OnDetaching() {
            base.OnDetaching();
            IsVisible = false;
        }
        protected override IPreviewModel GetPreviewModel() {
            return DocumentViewer.Model;
        }
        protected override void SetPreviewModel(IDocumentPreviewModel previewModel) {
            DocumentViewer.Model = previewModel;
        }
        protected override void SetCustomSettingsViewModel(object customSettingsViewModel) { }
    }
}

File: VB/DevExpress.OutlookInspiredApp.Wpf/Common/View/Services/ReportService.vb
Imports System
Imports System.ComponentModel
Imports System.Printing
Imports System.Windows
Imports DevExpress.DevAV.Common.ViewModel
Imports DevExpress.Mvvm.UI
Imports DevExpress.Xpf.Printing
Imports DevExpress.Xpf.Ribbon
Imports DevExpress.XtraReports

Namespace DevExpress.DevAV.Common.View
    Public MustInherit Class ReportServiceBase
        Inherits ServiceBase
        Implements IReportService

        Private Class PrintDirectXtraReportPreviewModel
            Inherits XtraReportPreviewModel

            Protected Overrides Sub PrintDirect(ByVal parameter As Object)
                If TypeOf parameter Is String Then
                    PrintDirect(DirectCast(parameter, String))
                ElseIf TypeOf parameter Is PrintQueue Then
                    PrintDirect(DirectCast(parameter, PrintQueue))
                Else
                    PrintDirect()
                End If
            End Sub
        End Class


        Private isVisible_Renamed As Boolean
        Private defaultReportInfo As IReportInfo
        Private reportInfo As IReportInfo
        Private actualReportInfo As IReportInfo

        Protected Property IsVisible() As Boolean
            Get
                Return isVisible_Renamed
            End Get
            Set(ByVal value As Boolean)
                isVisible_Renamed = value
                UpdateReport()
                If Not isVisible_Renamed Then
                    Me.reportInfo = Nothing
                End If
            End Set
        End Property
        Protected Overridable Sub SetDefaultReport(ByVal reportInfo As IReportInfo)
            Me.defaultReportInfo = reportInfo
            UpdateReport()
        End Sub
        Protected Overridable Sub ShowReport(ByVal reportInfo As IReportInfo)
            Me.reportInfo = reportInfo
            UpdateReport()
        End Sub
        Private Sub UpdateReport()
            UpdateReportCore(If(IsVisible, (If(reportInfo, defaultReportInfo)), Nothing))
        End Sub
        Protected Overridable Sub UpdateReportCore(ByVal actualReportInfo As IReportInfo)
            UnsubscribeFromParametersViewModel()
            Me.actualReportInfo = actualReportInfo
            SubscribeToParametersViewModel()
            If Me.actualReportInfo Is Nothing Then
                DestroyReport()
            Else
                CreateReport()
            End If
        End Sub
        Private Sub OnParametersViewModelPropertyChanged(ByVal sender As Object, ByVal e As PropertyChangedEventArgs)
            CreateReport()
        End Sub
        Private Sub CreateReport()
            Dim report As IReport = actualReportInfo.CreateReport()
            Dim model As PrintDirectXtraReportPreviewModel = TryCast(GetPreviewModel(), PrintDirectXtraReportPreviewModel)
            If model Is Nothing Then
                model = New PrintDirectXtraReportPreviewModel() With {.ZoomMode = New ZoomFitModeItem(ZoomFitMode.PageWidth)}
                SetPreviewModel(model)
            End If
            model.Report = report
            SetCustomSettingsViewModel(actualReportInfo.ParametersViewModel)
            report.PrintingSystemBase.ClearContent()
            report.CreateDocument(True)
        End Sub
        Private Sub DestroyReport()
            SetCustomSettingsViewModel(Nothing)
            SetPreviewModel(Nothing)
        End Sub
        Protected MustOverride Function GetPreviewModel() As IPreviewModel
        Protected MustOverride Sub SetPreviewModel(ByVal previewModel As IDocumentPreviewModel)
        Protected MustOverride Sub SetCustomSettingsViewModel(ByVal customSettingsViewModel As Object)
        Private ReadOnly Property ActualParametersViewModel() As Object
            Get
                Return If(Me.actualReportInfo Is Nothing, Nothing, Me.actualReportInfo.ParametersViewModel)
            End Get
        End Property
        Private Sub SubscribeToParametersViewModel()
            Dim parametersViewModel As INotifyPropertyChanged = TryCast(ActualParametersViewModel, INotifyPropertyChanged)
            If parametersViewModel IsNot Nothing Then
                AddHandler parametersViewModel.PropertyChanged, AddressOf OnParametersViewModelPropertyChanged
            End If
        End Sub
        Private Sub UnsubscribeFromParametersViewModel()
            Dim parametersViewModel As INotifyPropertyChanged = TryCast(ActualParametersViewModel, INotifyPropertyChanged)
            If parametersViewModel IsNot Nothing Then
                RemoveHandler parametersViewModel.PropertyChanged, AddressOf OnParametersViewModelPropertyChanged
            End If
        End Sub
        #Region "IReportService"
        Private Sub IReportService_SetDefaultReport(ByVal reportInfo As IReportInfo) Implements IReportService.SetDefaultReport
            SetDefaultReport(reportInfo)
        End Sub
        Private Sub IReportService_ShowReport(ByVal reportInfo As IReportInfo) Implements IReportService.ShowReport
            ShowReport(reportInfo)
        End Sub
        #End Region
    End Class
    Public Class BackstageDocumentPreviewReportService
        Inherits ReportServiceBase

        Public Shared ReadOnly BackstageViewIsOpenProperty As DependencyProperty = DependencyProperty.Register("BackstageViewIsOpen", GetType(Boolean), GetType(BackstageDocumentPreviewReportService), New PropertyMetadata(False, Sub(d, e) CType(d, BackstageDocumentPreviewReportService).OnBackstageViewIsOpenChanged()))
        Public Shared ReadOnly BackstageItemProperty As DependencyProperty = DependencyProperty.Register("BackstageItem", GetType(BackstageTabItem), GetType(BackstageDocumentPreviewReportService), New PropertyMetadata(Nothing, Sub(d, e) CType(d, BackstageDocumentPreviewReportService).OnBackstageItemChanged(CType(e.OldValue, BackstageItem), CType(e.NewValue, BackstageItem))))
        Public Shared ReadOnly BackstageDocumentPreviewProperty As DependencyProperty = DependencyProperty.Register("BackstageDocumentPreview", GetType(BackstageDocumentPreview), GetType(BackstageDocumentPreviewReportService))

        Public Property BackstageViewIsOpen() As Boolean
            Get
                Return CBool(GetValue(BackstageViewIsOpenProperty))
            End Get
            Set(ByVal value As Boolean)
                SetValue(BackstageViewIsOpenProperty, value)
            End Set
        End Property
        Public Property BackstageItem() As BackstageTabItem
            Get
                Return CType(GetValue(BackstageItemProperty), BackstageTabItem)
            End Get
            Set(ByVal value As BackstageTabItem)
                SetValue(BackstageItemProperty, value)
            End Set
        End Property
        Public Property BackstageDocumentPreview() As BackstageDocumentPreview
            Get
                Return CType(GetValue(BackstageDocumentPreviewProperty), BackstageDocumentPreview)
            End Get
            Set(ByVal value As BackstageDocumentPreview)
                SetValue(BackstageDocumentPreviewProperty, value)
            End Set
        End Property

        Protected Overrides Sub ShowReport(ByVal reportInfo As IReportInfo)
            MyBase.ShowReport(reportInfo)
            Dispatcher.BeginInvoke(CType(AddressOf OpenBackstageView, Action))
        End Sub
        Private Sub OpenBackstageView()
            BackstageItem.Backstage.SelectedTab = BackstageItem
            BackstageItem.Backstage.IsOpen = True
        End Sub
        Protected Overrides Sub UpdateReportCore(ByVal actualReportInfo As IReportInfo)
            MyBase.UpdateReportCore(actualReportInfo)
            BackstageItem.IsEnabled = actualReportInfo IsNot Nothing
        End Sub
        Private Sub OnBackstageViewIsOpenChanged()
            IsVisible = BackstageViewIsOpen
        End Sub
        Private Sub OnBackstageItemChanged(ByVal oldItem As BackstageItem, ByVal newItem As BackstageItem)
            If newItem IsNot Nothing Then
                newItem.IsEnabled = oldItem IsNot Nothing AndAlso oldItem.IsEnabled
            End If
        End Sub
        Protected Overrides Function GetPreviewModel() As IPreviewModel
            Return BackstageDocumentPreview.Model
        End Function
        Protected Overrides Sub SetPreviewModel(ByVal previewModel As IDocumentPreviewModel)
            BackstageDocumentPreview.Model = previewModel
        End Sub
        Protected Overrides Sub SetCustomSettingsViewModel(ByVal customSettingsViewModel As Object)
            BackstageDocumentPreview.CustomSettings = customSettingsViewModel
        End Sub
    End Class
    Public Class DocumentViewerReportService
        Inherits ReportServiceBase

        Private ReadOnly Property DocumentViewer() As DocumentViewer
            Get
                Return CType(AssociatedObject, DocumentViewer)
            End Get
        End Property

        Protected Overrides Sub OnAttached()
            MyBase.OnAttached()
            IsVisible = True
        End Sub
        Protected Overrides Sub OnDetaching()
            MyBase.OnDetaching()
            IsVisible = False
        End Sub
        Protected Overrides Function GetPreviewModel() As IPreviewModel
            Return DocumentViewer.Model
        End Function
        Protected Overrides Sub SetPreviewModel(ByVal previewModel As IDocumentPreviewModel)
            DocumentViewer.Model = previewModel
        End Sub
        Protected Overrides Sub SetCustomSettingsViewModel(ByVal customSettingsViewModel As Object)
        End Sub
    End Class
End Namespace
